xds/internal/client/client_test.go

```diff
@@ -121,53 +124,128 @@ func (c *testAPIClient) Close() {}
 	defer cleanup()
 
-	c, err := New(clientOpts(testXDSServer, false))
+	client, err := New(clientOpts(testXDSServer, false))
 	if err != nil {
 		t.Fatalf("failed to create client: %v", err)
 	}
-	defer c.Close()
+	defer client.Close()
 
-	v2Client := <-v2ClientCh
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	c, err := apiClientCh.Receive(ctx)
+	if err != nil {
+		t.Fatalf("timeout when waiting for API client to be created: %v", err)
+	}
+	apiClient := c.(*testAPIClient)
 
 	clusterUpdateCh := testutils.NewChannel()
 	firstTime := true
-	c.WatchCluster(testCDSName, func(update ClusterUpdate, err error) {
+	client.WatchCluster(testCDSName, func(update ClusterUpdate, err error) {
 		clusterUpdateCh.Send(clusterUpdateErr{u: update, err: err})
 		// Calls another watch inline, to ensure there's deadlock.
-		c.WatchCluster("another-random-name", func(ClusterUpdate, error) {})
+		client.WatchCluster("another-random-name", func(ClusterUpdate, error) {})
 
 		ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 		defer cancel()
-		if _, err := v2Client.addWatches[ClusterResource].Receive(ctx); firstTime && err != nil {
+		if _, err := apiClient.addWatches[ClusterResource].Receive(ctx); firstTime && err != nil {
 			t.Fatalf("want new watch to start, got error %v", err)
 		}
 		firstTime = false
 	})
 
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
-	if _, err := v2Client.addWatches[ClusterResource].Receive(ctx); err != nil {
+	if _, err := apiClient.addWatches[ClusterResource].Receive(ctx); err != nil {
 		t.Fatalf("want new watch to start, got error %v", err)
 	}
 
 	wantUpdate := ClusterUpdate{ServiceName: testEDSName}
-	v2Client.r.NewClusters(map[string]ClusterUpdate{
-		testCDSName: wantUpdate,
-	})
-
-	if u, err := clusterUpdateCh.Receive(ctx); err != nil || u != (clusterUpdateErr{wantUpdate, nil}) {
-		t.Errorf("unexpected clusterUpdate: %v, error receiving from channel: %v", u, err)
+	client.NewClusters(map[string]ClusterUpdate{testCDSName: wantUpdate})
+	if err := verifyClusterUpdate(clusterUpdateCh, wantUpdate); err != nil {
+		t.Fatal(err)
 	}
 
 	wantUpdate2 := ClusterUpdate{ServiceName: testEDSName + "2"}
-	v2Client.r.NewClusters(map[string]ClusterUpdate{
-		testCDSName: wantUpdate2,
-	})
+	client.NewClusters(map[string]ClusterUpdate{testCDSName: wantUpdate2})
+	if err := verifyClusterUpdate(clusterUpdateCh, wantUpdate2); err != nil {
+		t.Fatal(err)
+	}
+}
 
-	if u, err := clusterUpdateCh.Receive(ctx); err != nil || u != (clusterUpdateErr{wantUpdate2, nil}) {
-		t.Errorf("unexpected clusterUpdate: %v, error receiving from channel: %v", u, err)
+func verifyListenerUpdate(updateCh *testutils.Channel, wantUpdate ListenerUpdate) error {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
```

Comment:
It would be preferable to propagate context instead.

Reply:
Done.